src/dune/build.ml

```diff
@@ -180,101 +180,124 @@ let source_tree ~dir =
   in
   dirs_without_files >>> path_set paths >>> return paths
 
-let action ?dir ~targets action =
-  Targets (Path.Build.Set.of_list targets)
-  >>> return
-        ( match dir with
-        | None -> action
-        | Some dir -> Action.Chdir (dir, action) )
+module With_targets = struct
+  (* CR-soon amokhov: It seems to me that we can switch from [Path.Build.Set.t]
+     to [Path.Build.t list] since we should never have repeated targets. Or we
+     could use [union_exn] to check that targets are never repeated. *)
+  type nonrec 'a t =
+    { build : 'a t
+    ; targets : Path.Build.Set.t
+    }
+
+  let return x = { build = Pure x; targets = Path.Build.Set.empty }
+
+  let add t ~targets =
+    { build = t.build
+    ; targets = Path.Build.Set.union t.targets (Path.Build.Set.of_list targets)
+    }
+
+  let map { build; targets } ~f = { build = map build ~f; targets }
+
+  let map2 x y ~f =
+    { build = Map2 (f, x.build, y.build)
+    ; targets = Path.Build.Set.union x.targets y.targets
+    }
+
+  module O = struct
+    let ( >>> ) = map2 ~f:(fun () y -> y)
+
+    let ( and+ ) = map2 ~f:(fun x y -> (x, y))
+
+    let ( let+ ) a f = map ~f a
+  end
+
+  open O
+
+  let rec all xs =
+    match xs with
+    | [] -> return []
+    | x :: xs -> map2 ~f:List.cons x (all xs)
+
+  let write_file_dyn fn s =
+    add ~targets:[ fn ]
+      (let+ s = s in
+       Action.Write_file (fn, s))
+
+  let of_result_map res ~f ~targets =
+    add ~targets
+      ( match res with
+      | Ok x -> f x
+      | Error e ->
+        { build = Fail { fail = (fun () -> raise e) }
+        ; targets = Path.Build.Set.empty
+        } )
+end
+
+let add build ~targets : _ With_targets.t =
```

Comment:
Should this be called `Build.add_targets` or, perhaps, `Build.with_targets`? `add` makes me think of adding `Built.t` values together, or something like that.

Reply:
Yep, it used to be `Build.with_targets ~targets` but it felt a bit repetitive. I agree that `add` is an ambiguous name, so will bring `with_targets` back.